### ISSUE TITLE
Active subrange within composing range should be visually highlighted in Flutter

### ISSUE THREAD

**Issue by cbracken** [flutter/flutter#68547]
## Description

When inputting text in languages that use multi-step input (composing mode), there are two ranges involved:
1. The overall composing range
2. The active subrange within the composing range

This is common when inputting text using [ateji](https://en.wikipedia.org/wiki/Ateji) in Japanese, but occurs in general with longer composing ranges, person, and place names.

To support this, we'll need some visual highlighting of the text to make clear which range is the active subrange, as well as changes to the text input protocol so that the engine can specify this range to the framework. It seems unlikely to me that users will need to programmatically control the range via TextEditingController, so ideally the API surface should be as minimal as possible.

## Repro steps and expected behaviour

For example, when inputting the name 真惟佳 (Maika), one would go through the following steps:
1. Type the phonetic kana characters まいか.
2. Press shift-left arrow to restrict the active subrange within the composing range to the first character, ま.
3. Select the correct completion, then hit the right arrow to switch the active range to the remainder of the composing range. At this point the composing range reads 真いか and いか becomes the active subrange.
4. Press shift-left arrow to restrict the active range to the second character only, then select the correct completion, then hit right arrow to switch the active range to the final character. The composing region now read 真惟か and か becomes the active subrange.
5. Select the correct completion, then hit return to commit the composing region. The text 真惟佳 is now committed.

## Visual representation

The composing region and active subrange are each visually highlighted to the user in some platform-specific way. 

On macOS 11.0, Chrome, and Android, it is highlighted using a thicker underline with a small gap between it and the thinner composing region underline. In Linux GTK apps (and I believe older versions of macOS), the subrange is highlighted with a lighter colour than the standard selection colour.

macOS 11.0:
<img width="196" alt="image" src="https://user-images.githubusercontent.com/351029/96527441-e7b61000-1234-11eb-891d-e05b2fd88768.png">

iPadOS 14.0:
<img width="280" alt="image" src="https://user-images.githubusercontent.com/351029/96794656-718be780-13b3-11eb-8e7f-38244075acfe.png">

Chrome on Linux:
![image](https://user-images.githubusercontent.com/351029/96796384-8452ec00-13b4-11eb-9c3b-9ad138207b8a.png)

Linux GTK apps:
![image](https://user-images.githubusercontent.com/351029/96798259-5a4df980-13b5-11eb-8cff-1c33878ddec9.png)

**Comment by cbracken** [comment]
I believe this is a requirement for Desktop launch in CJK locales.

**Comment by HansMuller** [comment]
CC @LongCatIsLooong

**Comment by justinmc** [comment]
Is this something we're also missing on mobile?

**Comment by cbracken** [comment]
> Is this something we're also missing on mobile?

Yes. This is required on iOS devices using a physical keyboard. I've added a screenshot above.

**Comment by cbracken** [comment]
Example video on macOS to give a better idea:
https://user-images.githubusercontent.com/351029/110860551-c3a1ac80-8271-11eb-8877-2e1a05b6f82e.mov

**Comment by gnprice** [comment]
This issue also affects Android, even without a hardware keyboard.

With the Gboard Japanese keyboard, the expected behavior is just like in the original description above, except that instead of shift-left arrow, you use the software keyboard's left and right arrows. When a composing region is active, these arrows move the end of the active region, and are restricted to within the composing region.

In a native Android app, the highlighting is:
* The whole composing region is underlined.
* The active subregion is highlighted in an aqua-green hue. (Distinct from the selection color, which is more blue.)
* The rest of the composing region, if there is any, is highlighted with a lighter version of the same color.

Here's a screenshot after step 4(a) of the repro steps above: the middle of three characters is the active region, and the last two characters together are the composing region. This is on a Pixel device running Android 13. A screenshot of a selection is at the right for comparison.
![image](https://user-images.githubusercontent.com/28173/224526452-3bb278fd-b88f-4572-9e6e-b1c8409623dc.png) ![image](https://user-images.githubusercontent.com/28173/224527024-cf8d48de-51b3-4a9f-926e-f13c54ec1ed6.png)

(Oddly the caret is still painted at the end of the whole composing region. I checked multiple native apps and they all have this behavior. I don't know whether that's a design choice or a bug.)

The current actual behavior in Flutter main, on Android, is that the whole composing region is underlined and there is no highlighting or anything else to indicate the active subregion:
![image](https://user-images.githubusercontent.com/28173/224527157-248db260-74ad-473a-b6e4-8d4647707bf3.png)



<details><summary>Current behavior in more detail</summary>

In more detail, the current actual behavior is:
* If you ignore the visual feedback in the text field itself, and just focus on the keyboard, it actually works perfectly.

  But the feedback in the text field is confusing enough that many users may never discover that.
* The whole composing region is underlined, and the caret is painted at the end of the composing region, regardless of where the active subregion is.
  * Unless, that is, you shrink the active subregion all the way to be collapsed at the start of the composing region. In that case the caret is painted there at the start.

    Curiously the odd Android-native behavior mentioned above, where the caret is always painted at the end despite the active subregion, has this same exception: if you move all the way to the start of the composing region, then the caret is painted there instead of at the end.
* There is no indication of the active subregion at all. (Unless you collapse it all the way to the start.) So when you press the left arrow to try to shrink the active range, it looks in the text field like it didn't do anything.
  * (On the other hand if you look at the keyboard's suggestions, they will have updated exactly as they should. Also if you hit delete, or type more kana, the deletion or insertion happens correctly at the end of the active subregion, even though that's not where the caret appears to be.)

</details>

**Comment by gnprice** [comment]
A different issue that's also about styling the composing range, and highlighting vs. underlining:
* #9343

**Comment by cbracken** [comment]
Added a minor update to the styling issue for iOS 17/macOS 14 Sonoma. https://github.com/flutter/flutter/issues/9343

**Comment by LongCatIsLooong** [comment]
Reopening since https://github.com/flutter/engine/pull/49314 only fixes the issue on macOS (where we let the IME handle the text input key events). I haven't tried to reproduce the issue on iOS.
